yolo_to_yolo/data_transformers.py

```diff
@@ -47,3 +47,139 @@ def __call__(self, input_data: YoloImageData) -> Iterable[YoloImageData]:
+        x1_max, y1_max = bbox1.x + bbox1.w / 2, bbox1.y + bbox1.h / 2
+        x2_min, y2_min = bbox2.x - bbox2.w / 2, bbox2.y - bbox2.h / 2
+        x2_max, y2_max = bbox2.x + bbox2.w / 2, bbox2.y + bbox2.h / 2
+
+        intersect_x_min = max(x1_min, x2_min)
+        intersect_y_min = max(y1_min, y2_min)
+        intersect_x_max = min(x1_max, x2_max)
+        intersect_y_max = min(y1_max, y2_max)
+
+        if intersect_x_min < intersect_x_max and intersect_y_min < intersect_y_max:
+            intersection_area = (intersect_x_max - intersect_x_min) * (intersect_y_max - intersect_y_min)
+            bbox1_area = (x1_max - x1_min) * (y1_max - y1_min)
+            bbox2_area = (x2_max - x2_min) * (y2_max - y2_min)
+            iou = intersection_area / float(bbox1_area + bbox2_area - intersection_area)
+            return iou
+        return 0.0
+
+    def _get_combined_bbox(self, shape_bbox: YoloBbox, char_label: YoloLabel, img_width: int, img_height: int) -> tuple[int, int, int, int]:
+        """
+        Get the smallest bbox that contains both the shape and character bboxes
+        """
+        x_center, y_center, bbox_width, bbox_height = shape_bbox
+        x1 = int((x_center - bbox_width / 2) * img_width)
+        y1 = int((y_center - bbox_height / 2) * img_height)
+        x2 = int((x_center + bbox_width / 2) * img_width)
+        y2 = int((y_center + bbox_height / 2) * img_height)
+
+        if char_label:
+            char_bbox = char_label.location
+            char_x_center, char_y_center, char_bbox_width, char_bbox_height = char_bbox
+            char_x1 = int((char_x_center - char_bbox_width / 2) * img_width)
+            char_y1 = int((char_y_center - char_bbox_height / 2) * img_height)
+            char_x2 = int((char_x_center + char_bbox_width / 2) * img_width)
+            char_y2 = int((char_y_center + char_bbox_height / 2) * img_height)
+
+            x1 = min(x1, char_x1)
+            y1 = min(y1, char_y1)
+            x2 = max(x2, char_x2)
+            y2 = max(y2, char_y2)
+
+        return x1 + 1, y1 + 1, x2, y2
+    
+    def _adjust_bbox(self, bbox: YoloBbox, crop_x1: int, crop_y1: int, crop_width: int, crop_height: int, img_width: int, img_height: int) -> YoloBbox:
+        """
+        Adjust the bounding box coordinates to be relative to the cropped image.
+        (currently unused but we might need it later, it's called in a piece of commented code in __call__)
+        """
+        x_center, y_center, bbox_width, bbox_height = bbox
+        abs_x_center = x_center * img_width
+        abs_y_center = y_center * img_height
+        abs_width = bbox_width * img_width
+        abs_height = bbox_height * img_height
+
+        new_x_center = (abs_x_center - crop_x1) / crop_width
+        new_y_center = (abs_y_center - crop_y1) / crop_height
+        new_width = abs_width / crop_width
+        new_height = abs_height / crop_height
+
+        return YoloBbox(new_x_center, new_y_center, new_width, new_height)
```

Comment:
nitpick: can you put a newline at the end of the file? Leaving it out gives GitHub's file renderer an aneurysm.
![image](https://github.com/uci-uav-forge/imaging_training_2024/assets/48658337/8ea40989-d301-4648-8e84-e5f3d24801c2)